### PR TITLE
feat / Hanging orders in Avellaneda MM strategy

### PR DESCRIPTION
**Pull request by keithbaum**
**Before submitting this PR, please make sure**:

- [ ] Your code builds clean without any errors or warnings
- [ ] You are using approved title ("feat/", "fix/", "docs/", "refactor/")

**A description of the changes proposed in the pull request**:

### CODE REVIEW

**Comment by RC-13** [comment]
Getting some clock tick error related to restoring a hanging order. After one of the orders got filled and the other side gets cancelled, the equivalent hanging order fails to get restored with order_ID = `None`: 
image
```python
2021-06-03 14:41:27,004 - 13437 - hummingbot.core.clock - ERROR - Unexpected error running clock tick.
Traceback (most recent call last):
  File "hummingbot/core/clock.pyx", line 119, in hummingbot.core.clock.Clock.run_til
    child_iterator.c_tick(self._current_tick)
  File "hummingbot/strategy/avellaneda_market_making/avellaneda_market_making.pyx", line 465, in hummingbot.strategy.avellaneda_market_making.avellaneda_market_making.AvellanedaMarketMakingStrategy.c_tick
    self._hanging_orders_tracker.renew_hanging_orders_past_max_order_age()
  File "/Users/clint_g/Documents/Github-Repo/hb-general/hummingbot/strategy/hanging_orders_tracker.py", line 88, in renew_hanging_orders_past_max_order_age
    if order.age > max_age:
  File "/Users/clint_g/Documents/Github-Repo/hb-general/hummingbot/strategy/data_types.py", line 81, in age
    if self.creation_timestamp:
  File "/Users/clint_g/Documents/Github-Repo/hb-general/hummingbot/strategy/data_types.py", line 76, in creation_timestamp
    if "//" not in self.order_id:
TypeError: argument of type 'NoneType' is not iterable
```
[logs_conf_avellaneda_market_making_2.log](https://github.com/CoinAlpha/hummingbot/files/6589555/logs_conf_avellaneda_market_making_2.log)

**Comment by phbrgnomo** [comment]
Spotted an issue running the hanging orders as `volume_weighted`:

Captura de tela 2021-06-03 18 04 23

There should be only one hanging sell order(0.4807 FRONT at 0.00003228), but two duplicated orders were on the book for a some time (10-20 seconds).

Getting some disonnections with kucoin that might be the cause.

Captura de tela 2021-06-03 18 19 52

Also, got this error when exiting the bot.

Log
[logs_conf_avellaneda_market_making_4.log](https://github.com/CoinAlpha/hummingbot/files/6594424/logs_conf_avellaneda_market_making_4.log)

**Comment by RC-13** [comment]
Minor issue, `max_spread` currently doesn't have validation and accepts string values:
image

**Comment by keithbaum** [comment]
> Minor issue, `max_spread` currently doesn't have validation and accepts string values:
> image

Issue addressed in this mini PR https://github.com/CoinAlpha/hummingbot/pull/3511

**Comment by RC-13** [comment]
Confirming issues initially reported were fixed. Adding additional issue, the hanging order label suddenly gets removed and replaced as a second level sell order which should not be the case. The logs still recognize the other sell order as a hanging order.

https://user-images.githubusercontent.com/57189990/120969808-34873080-c79d-11eb-8a1a-ff2de54ec3a2.mov

**Comment by keithbaum** [comment]
Added some verbose messages to facilitate QA testing. This messages should be removed from version before release
image

**Comment by rapcmia** [comment]
Tested the bot using `hanging_orders_aggregation_type = no_aggregation`
- The bot started creating bid / ask orders, ok
- After the order getting filled (BUY), the bot returned the error `Unexpected error running clock tick.`

```
2021-06-07 17:42:03,992 - 8664 - hummingbot.core.clock - ERROR - Unexpected error running clock tick.
Traceback (most recent call last):
  File "hummingbot\core\clock.pyx", line 119, in hummingbot.core.clock.Clock.run_til
    child_iterator.c_tick(self._current_tick)
  File "hummingbot\strategy\avellaneda_market_making\avellaneda_market_making.pyx", line 466, in hummingbot.strategy.avellaneda_market_making.avellaneda_market_making.AvellanedaMarketMakingStrategy.c_tick
    self.c_cancel_active_orders(proposal)
  File "hummingbot\strategy\avellaneda_market_making\avellaneda_market_making.pyx", line 1049, in hummingbot.strategy.avellaneda_market_making.avellaneda_market_making.AvellanedaMarketMakingStrategy.c_cancel_active_orders
    if not self._hanging_orders_tracker.is_order_to_be_added_to_hanging_orders(order):
  File "C:\Users\Ralph\github\3463\hummingbot\strategy\hanging_orders_tracker.py", line 145, in is_order_to_be_added_to_hanging_orders
    return any(o.client_id == order.client_order_id
  File "C:\Users\Ralph\github\3463\hummingbot\strategy\hanging_orders_tracker.py", line 145, in <genexpr>
    return any(o.client_id == order.client_order_id
AttributeError: 'HangingOrder' object has no attribute 'client_id'
```

- Stop/Exit the bot, relaunch, and start again. The bot created bid/ask orders ok
- Then the bot had an order filled again and return the same error after:

```
2021-06-07 18:22:55,019 - 4432 - hummingbot.strategy.avellaneda_market_making.avellaneda_market_making - INFO - (FRONT-BTC) Maker buy order x-XEKWYICX-BFTBC1623061355000211 (11.00000000 FRONT @ 0.00002962 BTC) has been completely filled. [clock=2021-06-07 10:22:54+00:00]
```

no_aggregate after filled order event

Tested the bot using `hanging_orders_aggregation_type = volume_weighted`
- The bot created bid/ask orders ok
- The bot had 2 filled orders but did not show on the logs that its accumulating hanging_orders
```
x-XEKWYICX-BFTBC1623061900008536 (11.00000000 FRONT @ 0.00002964 BTC) has been completely filled. [clock=2021-06-07 10:32:09+00:00]
2021-06-07 18:32:14,006 - 9780 - hummingbot.strategy.avellaneda_market_making.avellaneda_market_making - INFO - (FRONT-BTC) Cancelling the limit order x-XEKWYICX-SFTBC1623061900008537. [clock=2021-06-07 10:32:14+00:00]
2021-06-07 18:32:14,101 - 9780 - hummingbot.connector.exchange.binance.binance_exchange - INFO - Successfully cancelled order x-XEKWYICX-SFTBC1623061900008537.
2021-06-07 18:32:14,102 - 9780 - hummingbot.core.event.event_reporter - EVENT_LOG - {"timestamp": 1623061934.0, "order_id": "x-XEKWYICX-SFTBC1623061900008537", "exchange_order_id": null, "event_name": "OrderCancelledEvent", "event_source": "binance"}
2021-06-07 18:32:15,010 - 9780 - hummingbot.strategy.avellaneda_market_making.avellaneda_market_making - INFO - (FRONT-BTC) Creating 1 bid orders at (Size, Price): ['11 FRONT, 0.00002961 BTC']
2021-06-07 18:32:15,011 - 9780 - hummingbot.strategy.avellaneda_market_making.avellaneda_market_making - INFO - (FRONT-BTC) Creating 1 ask orders at (Size, Price): ['6.4 FRONT, 0.00002975 BTC']
2021-06-07 18:32:15,105 - 9780 - hummingbot.connector.exchange.binance.binance_exchange - INFO - Created LIMIT_MAKER BUY order x-XEKWYICX-BFTBC1623061935011558 for 11.00000000 FRONT-BTC.
2021-06-07 18:32:15,105 - 9780 - hummingbot.core.event.event_reporter - EVENT_LOG - {"timestamp": 1623061935.0, "type": "OrderType.LIMIT_MAKER", "trading_pair": "FRONT-BTC", "amount": "11.00000000", "price": "0.00002961", "order_id": "x-XEKWYICX-BFTBC1623061935011558", "exchange_order_id": "18437729", "leverage": 1, "position": "NILL", "event_name": "BuyOrderCreatedEvent", "event_source": "binance"}
2021-06-07 18:32:15,611 - 9780 - hummingbot.connector.exchange.binance.binance_exchange - INFO - Created LIMIT_MAKER SELL order x-XEKWYICX-SFTBC1623061935011559 for 6.40000000 FRONT-BTC.
2021-06-07 18:32:15,611 - 9780 - hummingbot.core.event.event_reporter - EVENT_LOG - {"timestamp": 1623061935.0, "type": "OrderType.LIMIT_MAKER", "trading_pair": "FRONT-BTC", "amount": "6.40000000", "price": "0.00002975", "order_id": "x-XEKWYICX-SFTBC1623061935011559", "exchange_order_id": "18437730", "leverage": 1, "position": "NILL", "event_name": "SellOrderCreatedEvent", "event_source": "binance"}
2021-06-07 18:33:15,017 - 9780 - hummingbot.strategy.avellaneda_market_making.avellaneda_market_making - INFO - (FRONT-BTC) Cancelling the limit order x-XEKWYICX-BFTBC1623061935011558. [clock=2021-06-07 10:33:15+00:00]
2021-06-07 18:33:15,017 - 9780 - hummingbot.strategy.avellaneda_market_making.avellaneda_market_making - INFO - (FRONT-BTC) Cancelling the limit order x-XEKWYICX-SFTBC1623061935011559. [clock=2021-06-07 10:33:15+00:00]
2021-06-07 18:33:15,113 - 9780 - hummingbot.connector.exchange.binance.binance_exchange - INFO - Successfully cancelled order x-XEKWYICX-BFTBC1623061935011558.
2021-06-07 18:33:15,114 - 9780 - hummingbot.core.event.event_reporter - EVENT_LOG - {"timestamp": 1623061995.0, "order_id": "x-XEKWYICX-BFTBC1623061935011558", "exchange_order_id": null, "event_name": "OrderCancelledEvent", "event_source": "binance"}
2021-06-07 18:33:15,616 - 9780 - hummingbot.connector.exchange.binance.binance_exchange - INFO - Successfully cancelled order x-XEKWYICX-SFTBC1623061935011559.
2021-06-07 18:33:15,616 - 9780 - hummingbot.core.event.event_reporter - EVENT_LOG - {"timestamp": 1623061995.0, "order_id": "x-XEKWYICX-SFTBC1623061935011559", "exchange_order_id": null, "event_name": "OrderCancelledEvent", "event_source": "binance"}
2021-06-07 18:33:16,000 - 9780 - hummingbot.strategy.avellaneda_market_making.avellaneda_market_making - INFO - (FRONT-BTC) Creating 1 bid orders at (Size, Price): ['11 FRONT, 0.00002965 BTC']
2021-06-07 18:33:16,001 - 9780 - hummingbot.strategy.avellaneda_market_making.avellaneda_market_making - INFO - (FRONT-BTC) Creating 1 ask orders at (Size, Price): ['6.4 FRONT, 0.0000297 BTC']
```

- Did the same steps of stop/exit and start the bot again and it created bid/ask orders ok
- After an order getting filled it return the same traceback error for the ` Unexpected error running clock tick.`
volume_weighted after filled order


Attached logs:
[logs_ave-3463_volume_weighted.txt](https://github.com/CoinAlpha/hummingbot/files/6608975/logs_ave-3463_volume_weighted.txt)
[logs_ave-3463_no_aggregate.txt](https://github.com/CoinAlpha/hummingbot/files/6608977/logs_ave-3463_no_aggregate.txt)

**Comment by keithbaum** [comment]
> Tested the bot using `hanging_orders_aggregation_type = no_aggregation`
> 
> * The bot started creating bid / ask orders, ok
> * After the order getting filled (BUY), the bot returned the error `Unexpected error running clock tick.`
> 
> ```
> 2021-06-07 17:42:03,992 - 8664 - hummingbot.core.clock - ERROR - Unexpected error running clock tick.
> Traceback (most recent call last):
>   File "hummingbot\core\clock.pyx", line 119, in hummingbot.core.clock.Clock.run_til
>     child_iterator.c_tick(self._current_tick)
>   File "hummingbot\strategy\avellaneda_market_making\avellaneda_market_making.pyx", line 466, in hummingbot.strategy.avellaneda_market_making.avellaneda_market_making.AvellanedaMarketMakingStrategy.c_tick
>     self.c_cancel_active_orders(proposal)
>   File "hummingbot\strategy\avellaneda_market_making\avellaneda_market_making.pyx", line 1049, in hummingbot.strategy.avellaneda_market_making.avellaneda_market_making.AvellanedaMarketMakingStrategy.c_cancel_active_orders
>     if not self._hanging_orders_tracker.is_order_to_be_added_to_hanging_orders(order):
>   File "C:\Users\Ralph\github\3463\hummingbot\strategy\hanging_orders_tracker.py", line 145, in is_order_to_be_added_to_hanging_orders
>     return any(o.client_id == order.client_order_id
>   File "C:\Users\Ralph\github\3463\hummingbot\strategy\hanging_orders_tracker.py", line 145, in <genexpr>
>     return any(o.client_id == order.client_order_id
> AttributeError: 'HangingOrder' object has no attribute 'client_id'
> ```
> 
>  * Stop/Exit the bot, relaunch, and start again. The bot created bid/ask orders ok
> * Then the bot had an order filled again and return the same error after:
> 
> ```
> 2021-06-07 18:22:55,019 - 4432 - hummingbot.strategy.avellaneda_market_making.avellaneda_market_making - INFO - (FRONT-BTC) Maker buy order x-XEKWYICX-BFTBC1623061355000211 (11.00000000 FRONT @ 0.00002962 BTC) has been completely filled. [clock=2021-06-07 10:22:54+00:00]
> ```
> 
> no_aggregate after filled order event
> 
> Tested the bot using `hanging_orders_aggregation_type = volume_weighted`
> 
> * The bot created bid/ask orders ok
> * The bot had 2 filled orders but did not show on the logs that its accumulating hanging_orders
> 
> ```
> x-XEKWYICX-BFTBC1623061900008536 (11.00000000 FRONT @ 0.00002964 BTC) has been completely filled. [clock=2021-06-07 10:32:09+00:00]
> 2021-06-07 18:32:14,006 - 9780 - hummingbot.strategy.avellaneda_market_making.avellaneda_market_making - INFO - (FRONT-BTC) Cancelling the limit order x-XEKWYICX-SFTBC1623061900008537. [clock=2021-06-07 10:32:14+00:00]
> 2021-06-07 18:32:14,101 - 9780 - hummingbot.connector.exchange.binance.binance_exchange - INFO - Successfully cancelled order x-XEKWYICX-SFTBC1623061900008537.
> 2021-06-07 18:32:14,102 - 9780 - hummingbot.core.event.event_reporter - EVENT_LOG - {"timestamp": 1623061934.0, "order_id": "x-XEKWYICX-SFTBC1623061900008537", "exchange_order_id": null, "event_name": "OrderCancelledEvent", "event_source": "binance"}
> 2021-06-07 18:32:15,010 - 9780 - hummingbot.strategy.avellaneda_market_making.avellaneda_market_making - INFO - (FRONT-BTC) Creating 1 bid orders at (Size, Price): ['11 FRONT, 0.00002961 BTC']
> 2021-06-07 18:32:15,011 - 9780 - hummingbot.strategy.avellaneda_market_making.avellaneda_market_making - INFO - (FRONT-BTC) Creating 1 ask orders at (Size, Price): ['6.4 FRONT, 0.00002975 BTC']
> 2021-06-07 18:32:15,105 - 9780 - hummingbot.connector.exchange.binance.binance_exchange - INFO - Created LIMIT_MAKER BUY order x-XEKWYICX-BFTBC1623061935011558 for 11.00000000 FRONT-BTC.
> 2021-06-07 18:32:15,105 - 9780 - hummingbot.core.event.event_reporter - EVENT_LOG - {"timestamp": 1623061935.0, "type": "OrderType.LIMIT_MAKER", "trading_pair": "FRONT-BTC", "amount": "11.00000000", "price": "0.00002961", "order_id": "x-XEKWYICX-BFTBC1623061935011558", "exchange_order_id": "18437729", "leverage": 1, "position": "NILL", "event_name": "BuyOrderCreatedEvent", "event_source": "binance"}
> 2021-06-07 18:32:15,611 - 9780 - hummingbot.connector.exchange.binance.binance_exchange - INFO - Created LIMIT_MAKER SELL order x-XEKWYICX-SFTBC1623061935011559 for 6.40000000 FRONT-BTC.
> 2021-06-07 18:32:15,611 - 9780 - hummingbot.core.event.event_reporter - EVENT_LOG - {"timestamp": 1623061935.0, "type": "OrderType.LIMIT_MAKER", "trading_pair": "FRONT-BTC", "amount": "6.40000000", "price": "0.00002975", "order_id": "x-XEKWYICX-SFTBC1623061935011559", "exchange_order_id": "18437730", "leverage": 1, "position": "NILL", "event_name": "SellOrderCreatedEvent", "event_source": "binance"}
> 2021-06-07 18:33:15,017 - 9780 - hummingbot.strategy.avellaneda_market_making.avellaneda_market_making - INFO - (FRONT-BTC) Cancelling the limit order x-XEKWYICX-BFTBC1623061935011558. [clock=2021-06-07 10:33:15+00:00]
> 2021-06-07 18:33:15,017 - 9780 - hummingbot.strategy.avellaneda_market_making.avellaneda_market_making - INFO - (FRONT-BTC) Cancelling the limit order x-XEKWYICX-SFTBC1623061935011559. [clock=2021-06-07 10:33:15+00:00]
> 2021-06-07 18:33:15,113 - 9780 - hummingbot.connector.exchange.binance.binance_exchange - INFO - Successfully cancelled order x-XEKWYICX-BFTBC1623061935011558.
> 2021-06-07 18:33:15,114 - 9780 - hummingbot.core.event.event_reporter - EVENT_LOG - {"timestamp": 1623061995.0, "order_id": "x-XEKWYICX-BFTBC1623061935011558", "exchange_order_id": null, "event_name": "OrderCancelledEvent", "event_source": "binance"}
> 2021-06-07 18:33:15,616 - 9780 - hummingbot.connector.exchange.binance.binance_exchange - INFO - Successfully cancelled order x-XEKWYICX-SFTBC1623061935011559.
> 2021-06-07 18:33:15,616 - 9780 - hummingbot.core.event.event_reporter - EVENT_LOG - {"timestamp": 1623061995.0, "order_id": "x-XEKWYICX-SFTBC1623061935011559", "exchange_order_id": null, "event_name": "OrderCancelledEvent", "event_source": "binance"}
> 2021-06-07 18:33:16,000 - 9780 - hummingbot.strategy.avellaneda_market_making.avellaneda_market_making - INFO - (FRONT-BTC) Creating 1 bid orders at (Size, Price): ['11 FRONT, 0.00002965 BTC']
> 2021-06-07 18:33:16,001 - 9780 - hummingbot.strategy.avellaneda_market_making.avellaneda_market_making - INFO - (FRONT-BTC) Creating 1 ask orders at (Size, Price): ['6.4 FRONT, 0.0000297 BTC']
> ```
> 
>  * Did the same steps of stop/exit and start the bot again and it created bid/ask orders ok
> * After an order getting filled it return the same traceback error for the ` Unexpected error running clock tick.`
>   volume_weighted after filled order
> 
> Attached logs:
> [logs_ave-3463_volume_weighted.txt](https://github.com/CoinAlpha/hummingbot/files/6608975/logs_ave-3463_volume_weighted.txt)
> [logs_ave-3463_no_aggregate.txt](https://github.com/CoinAlpha/hummingbot/files/6608977/logs_ave-3463_no_aggregate.txt)

Issue should be addressed in latest commit

File: hummingbot/templates/conf_avellaneda_market_making_strategy_TEMPLATE.yml
Comment (aarmoa) [inline hummingbot/templates/conf_avellaneda_market_making_strategy_TEMPLATE.yml:47]: typo in aggregation (Select way of aggregating)

**Review by aarmoa — COMMENTED** [review]
I have added some comments. I will wait for Nico's feedback on them before the last review.

File: hummingbot/strategy/hanging_orders_tracker.py
Comment (keithbaum) [reply hummingbot/strategy/hanging_orders_tracker.py]: Simplicity = Elegance :)

**Review by aarmoa — APPROVED** [review]
I agree with the changes to add hanging orders functionality to the strategy

File: hummingbot/strategy/hanging_orders_tracker.py
Comment (keithbaum) [reply hummingbot/strategy/hanging_orders_tracker.py:66]: This was discussed with Abel, and we decided is best to keep default arguments as None

**Comment by keithbaum** [comment]
Committed some minor changes after @aarmoa review. Think this should be the final version to review

**Review by aarmoa — DISMISSED** [review]
Latest fixes look good to me

**Comment by keithbaum** [comment]
Orders' aggregation as per request by @phbrgnomo today. It should now be good to test. Thanks

**Review by aarmoa — APPROVED** [review]
The change to handle hanging orders for buys and sells in different groups looks good to me

**Comment by phbrgnomo** [comment]
Before merge, we must change the `hanging_orders_aggregation_type`. They are too "wordy" at the moment.

Final choices should be:

`no_aggregation`
`volume_weighted`
`time_weighted`
`distance_weighted`

**Review by RC-13 — APPROVED** [review]
Tested and confirmed working by QA.

**Pull request merged**
